Rework the relationship between workflows and entities.

ttung opened the pull request:
### Description
The key changes is that only one workflow is considered to ever be responsible for creating an entity.  If more than one workflow contributes to an entity, we should have separate entities to represent them.

Added a method to easily traverse between entities and its parents (entities that feed into the workflow that creates the entity in question) and children (entities that result from workflows that consume the entity in question).

### Test plan
used in next PR

phoenixAja commented on src/py/aspen/database/models/entity.py:
```diff
@@ -1,12 +1,31 @@
 from .base import base, idbase
 from .enum import Enum
 
+if TYPE_CHECKING:
+    from .workflow import Workflow
```
why do you need this part?

ttung replied:
I use it in type annotations, for example [here](https://github.com/chanzuckerberg/aspen/pull/81/files#diff-c25c7ebf669612dcdc0aa054e26ff3d103f4b125ab2b05a48a63a0691357e552R67).  If I don't gate it with `if TYPE_CHECKING`, however, I get recursive imports.

ttung replied:
to be specific, mypy is two-pass, normal python execution is one-pass.

Commit:
Rework the relationship between workflows and entities.

### Description
The key changes is that only one workflow is considered to ever be responsible for creating an entity.  If more than one workflow contributes to an entity, we should have separate entities to represent them.

Added a method to easily traverse between entities and its parents (entities that feed into the workflow that creates the entity in question) and children (entities that result from workflows that consume the entity in question).

### Test plan
used in next PR
